File: 2024/python/day13/solution.py
def is_robot_back(moves: list[str]) -> bool | list[int]:
    modifiers = ["*", "!", "?"]
    commands = {
        "R": (0, 1, "L"),
        "L": (0, -1, "R"),
        "U": (1, 1, "D"),
        "D": (1, -1, "U"),
    }
    done = []
    modifier = None
    position = [0, 0]
    for i, mov in enumerate(moves):
        if mov in modifiers:
            modifier = mov
            continue
        command = mov
        if modifier:
            if modifier == "*":
                command = command * 2
            if modifier == "!":
                command = commands[mov][2]
            if modifier == "?":
                if command in done:
                    command = None
            modifier = None
        if command:
            for movement in command:
                index, increment, _ = commands[movement]
                position[index] = position[index] + increment
            done.append(command[0])
    if position == [0, 0]:
        return True
    return position

File: 2024/js/day13/solution.js
function isRobotBack(moves) {
  const modifiers = ["*", "!", "?"];
  const commands = {
    R: [0, 1, "L"],
    L: [0, -1, "R"],
    U: [1, 1, "D"],
    D: [1, -1, "U"],
  };

  const done = [];
  let modifier = null;
  const position = [0, 0];

  for (let i = 0; i < moves.length; i++) {
    const mov = moves[i];

    if (modifiers.includes(mov)) {
      modifier = mov;
      continue;
    }

    let command = mov;

    if (modifier) {
      if (modifier === "*") {
        command = command.repeat(2);
      }
      if (modifier === "!") {
        command = commands[mov][2];
      }
      if (modifier === "?") {
        if (done.includes(command)) {
          command = null;
        }
      }
      modifier = null;
    }

    if (command) {
      for (const movement of command) {
        const [index, increment] = commands[movement];
        position[index] = position[index] + increment;
      }
      done.push(command[0]);
    }
  }

  return position[0] === 0 && position[1] === 0 ? true : position;
}

